clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+      : SyclKernelFieldHandler(S), KernelObj(KernelObj), ArgTys(ArgTys),
+        Params(Params) {}
+
+  void handleSyclAccessorType(const FieldDecl *FD, QualType ArgTy) final {
+    handleSpecialType(FD, ArgTy);
+  }
+
+  void handleStructType(const FieldDecl *FD, QualType ArgTy) override {
+    addParam(FD, ArgTy);
+
+    // Create descriptors for each accessor field in the class or struct
+    const auto *Wrapper = ArgTy->getAsCXXRecordDecl();
+    VisitKernelFields(Wrapper->fields(), *this);
+  }
+};
+
+// A type to Create and own the FunctionDecl for the kernel.
+class SyclKernelDeclCreator : public SyclKernelDeclBase {
+  // TODO: rather than this, should we consider a 'commit' function that
+  // finalizes under success only?
+  SyclKernelFieldChecker &ArgChecker;
+  Sema::ContextRAII FuncContext;
+  // Yes, the list of Parameters contains this info, but we use it often enough
+  // we shouldn't be recreating it constantly.  QualTypes are cheap anyway.
+  llvm::SmallVector<QualType, 8> ArgTys;
+  llvm::SmallVector<ParmVarDecl *, 8> Params;
+
+  static void setKernelImplicitAttrs(ASTContext &Context, FunctionDecl *FD,
+                                     StringRef Name) {
+    // Set implict attributes.
+    FD->addAttr(OpenCLKernelAttr::CreateImplicit(Context));
+    FD->addAttr(AsmLabelAttr::CreateImplicit(Context, Name));
+    FD->addAttr(ArtificialAttr::CreateImplicit(Context));
+  }
+
+  static FunctionDecl *initKernelObj(ASTContext &Ctx, StringRef Name,
+                                     SourceLocation Loc, bool IsInline) {
+    // Create this with no prototype, and we can fix this up after we've seen
+    // all the params.
+    FunctionProtoType::ExtProtoInfo Info(CC_OpenCLKernel);
+    QualType FuncType = Ctx.getFunctionType(Ctx.VoidTy, {}, Info);
+
+    FunctionDecl *FD = FunctionDecl::Create(
+        Ctx, Ctx.getTranslationUnitDecl(), Loc, Loc, &Ctx.Idents.get(Name),
+        FuncType, Ctx.getTrivialTypeSourceInfo(Ctx.VoidTy), SC_None);
+    FD->setImplicitlyInline(IsInline);
+    setKernelImplicitAttrs(Ctx, FD, Name);
+
+    // Add kernel to translation unit to see it in AST-dump.
+    Ctx.getTranslationUnitDecl()->addDecl(FD);
+    return FD;
+  }
+
+public:
+  SyclKernelDeclCreator(Sema &S, SyclKernelFieldChecker &ArgChecker,
+                        StringRef Name, SourceLocation Loc, bool IsInline)
+      : SyclKernelDeclBase(
+            S, initKernelObj(S.getASTContext(), Name, Loc, IsInline), ArgTys,
+            Params),
+        ArgChecker(ArgChecker), FuncContext(SemaRef, KernelObj) {}
+
+  ~SyclKernelDeclCreator() {
+    KernelObj->setParams(Params);
+
+    if (ArgChecker.isValid())
+      SemaRef.addSyclDeviceDecl(KernelObj);
+  }
+
+  void handleSyclSamplerType(const FieldDecl *FD, QualType ArgTy) final {
+    handleSpecialType(FD, ArgTy);
+  }
+
+  void handlePointerType(const FieldDecl *FD, QualType ArgTy) final {
+    // TODO: Can we document what the heck this is doing?!
```

Comment:
Yep, that would be nice. But it's not so hard to understand what is happening here looking into the code.
This is just handling of USM pointers. I haven't seen any doc about implementation details of USM, only specs (https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions/USM).
Basically USM allows to use raw pointers instead of buffers/accessors, but these pointers point to the specially allocated memory. For any pointer we add some argument with the same type but global address space, because OpenCL requires it.
The problem here that we don't know how to differentiate USM pointers from any others, so we stupidly consider them all as USM pointers. This can lead to some problems, I guess.

Reply:
I've read both the code and your post and don't have a good hold of it other than it seems to be just creating a pointer to the same type, except global.  Patches that do a good job explaining the details would be welcome :)

Reply:
Hi. Sorry.

At least when I did this, kernel args to OpenCL that are pointers had to be in the global address space.

I make no claims to having done the right thing, only that I did something that got pointers to work.

And yes - I don't see a way to disambiguate 'normal' C++ pointers from USM pointers since USM is very specifically trying to avoid modifying the type system.
Previously, the compiler was rejecting all programs with pointers, so things that should work wouldn't.
We had to make it so any pointer works, which means that legal things now work, but illegal things will crash.  I still view that as an improvement.